Added How To page for Provision an Ubuntu virtual machine running SQL Server in Azure and  AWS Pro activation (#210)
canonical/ubuntu-cloud-docs: aws/aws-how-to/instances/upgrade-to-ubuntu-pro-at-scale-using-tokens-with-ssm.rst

Upgrade to Ubuntu Pro at scale using tokens with SSM
====================================================

Overview

> ---------

Reviewer: RST is very picky about the length of these underlines matching the length of the word they underline. Please fix these so the CI jobs can pass.

> ---------

Reviewer: ```suggestion
---------
```
```suggestion
--------
```

Ubuntu Pro expands Ubuntu LTS’ 5 years of coverage up to ten-year of security coverage with an additional support of 23,000 packages beyond the main operating system and it is free for use on up to 5 machines!

There are three ways to get Ubuntu Pro on AWS:

    1. Launching an Ubuntu Pro machine from the AWS Marketplace, which is ready to use through a PAYG charging model.
    2. Attaching an Ubuntu Pro token to any existing Ubuntu LTS server
    3. Upgrading to Ubuntu Pro using AWS License Manager

This how-to will cover the second case: Activating Pro via token using AWS Systems Manager.

There are many ways to automate actions on AWS Systems Manager but for this how-to, we are going to use SSM Documents and the Run Command functionality.

Prerequisites
-------------

> - An AWS account

Author: I don't know the answer to this. I am not the original author of the tutorial, I've just adapted it for style as requested in Issue #48 "Public Cloud: Write a how-to guide for Azure based on an existing tutorial https://github.com/canonical/ubuntu-cloud-docs/pull/48" from Open Documentation Academy.

I'm happy to make adjustments here if you can direct me to where I could find the answer to this sort of question.

> - An AWS account

Reviewer: I would at least add something like "an AWS account with write access to SSM and the ability to connect to VMs [via Fleet Manager]". It seems like those are the two main pieces of functionality required at the AWS level.

- An Ubuntu One account (you can create one on `Ubuntu Pro | Ubuntu <https://ubuntu.com/pro>`_)
- Machines with SSM agent installed (available by default on Ubuntu 16.04 onwards). If the agent is properly configured, the servers will appear as Managed Nodes on AWS SSM Fleet Manager (see the reference links for more information)
- Basic understanding about writing and running SSM Documents (see the reference links for more information)

1. Get a token

> ------------

Reviewer: ```suggestion
------------
```
```suggestion
--------------
```

> To get an Ubuntu Pro token, you need to create an account in `Ubuntu Pro | Ubuntu <https://ubuntu.com/pro>`_ to see your subscriptions and get your tokens. Anyone can use Ubuntu Pro for free on up to 5 machines, even for production workloads.

Author: I don't know the exact definition for machines here. This phrasing is from the original tutorial and is also the phrasing used in announcements such as this: https://ubuntu.com/blog/ubuntu-pro-beta-release

Log-in by clicking on *Your subscriptions*. You will be redirected to the following screen where the token is exposed as shown in the orange ellipse:

.. image:: upgrade-to-ubuntu-pro-at-scale-using-tokens-with-ssm-images/0_token_screen.png
   :align: center

2. Create a SSM Document to attach the token

> ------------------------------------------

Reviewer: ```suggestion
------------------------------------------
```
```suggestion
--------------------------------------------
```

The SSM Document is just a JSON or YAML file containing the script in steps or sections.

    Note:
    As the token is personal, it is not recommended to hard-code it into the code but add it as a user parameter.

This is the full code:

.. code-block::

    ---
    schemaVersion: '2.2'
    description: "Enable Pro services using a Pro token"
    parameters:
    token:
        type: String
    mainSteps:
    - action: aws:runShellScript
    name: update
    inputs:
        runCommand:
        - "sudo apt-get update && sudo apt-get upgrade -y"
    - action: aws:runShellScript
    name: attachToken
    inputs:
        runCommand:
        - "sudo pro attach {{ token }}"
    - action: aws:runShellScript
    name: updateWithESM
    inputs:
        runCommand:
        - "sudo apt-get update && sudo apt-get upgrade -y"

> The token will be requested to the user when running this document, then it will execute the rest of the code:

Reviewer: This sentence needs re-wording. "The token will be requested to the user" is unclear. Also unclear is what is meant by "it" in "then it will execute the rest of the code."

>     1. Run an update (to make sure the pro agent is at the latest version)

Reviewer: What kind of update? APT update? Can you add an example command here?

Author: This is handled by the SSM document itself. I've rephrased to clarify that.

>     2. Attach the token

Reviewer: How does one attach the token?

    3. Update again in case you have installed software that could get security updates from Pro repositories.

No reboot is needed.

3. Run the SSM script

> ------------------

Reviewer: ```suggestion
------------------
```
```suggestion
---------------------
```

> The SSM script can be run either from the Fleet Manager or with the SSM Run Command. More information can be found in `this tutorial <https://ubuntu.com/tutorials/how-to-create-ssm-documents-and-use-them-to-install-packages-massively-with-run-command#1-overview>`

Reviewer: This might be a good place to describe what "the SSM script" is

Author: The SSM script is described in the previous step.

> You don’t need to restart the machines after applying the token. If it fails, check the SSM format of your document, see if there are any typos and try again. If this does not work check the output generated by SSM and see for any configuration issue (do your machines have access to update and upgrade packages? Pro repositories are available on different endpoints. Check the references at the end of the tutorial).

Reviewer: ```suggestion
You don’t need to restart the machines after applying the token. If the SSM script fails, check the SSM format of your document, see if there are any typos and try again. If this does not work check the output generated by SSM and look for any configuration issue (do your machines have access to update and upgrade packages? Pro repositories are available on different endpoints [[than what?]]. Check the references at the end of the tutorial).
```

What endpoints are Pro repositories available at? Are these static, and if so, can they be listed here?

.. image:: upgrade-to-ubuntu-pro-at-scale-using-tokens-with-ssm-images/1_run_command_screen.png
   :align: center

> Check the status

Reviewer: ```suggestion
Check the Pro status of your machine
```

----------------

You can always check the status of the subscription by running: 

.. code-block::
    
    sudo pro status 

> and see entitlements and enabled services.

Reviewer: ```suggestion
This will display your entitlements and enabled services.
```

> ----------------

Reviewer: ```suggestion
----------------
```
```suggestion
---------------------------------------
```
